tokio/src/util/linked_list.rs

```diff
@@ -4,95 +4,121 @@
-    _pin: PhantomPinned,
+    next: Option<NonNull<T>>,
 }
 
-unsafe impl<T: Send> Send for Entry<T> {}
-unsafe impl<T: Sync> Sync for Entry<T> {}
+unsafe impl<T: Send> Send for Pointers<T> {}
+unsafe impl<T: Sync> Sync for Pointers<T> {}
+
+// ===== impl LinkedList =====
 
-impl<T> LinkedList<T> {
+impl<T: Link> LinkedList<T> {
     /// Creates an empty linked list
-    pub(crate) fn new() -> Self {
+    pub(crate) fn new() -> LinkedList<T> {
         LinkedList {
             head: None,
             tail: None,
+            _p: PhantomData,
         }
     }
 
-    /// Adds an item to the back of the linked list.
-    ///
-    /// # Safety
-    ///
-    /// The function is only safe as long as valid pointers are stored inside
-    /// the linked list.
-    pub(crate) unsafe fn push_front(&mut self, entry: Pin<&mut Entry<T>>) {
-        let mut entry: NonNull<Entry<T>> = entry.get_unchecked_mut().into();
+    /// Adds an element first in the list.
+    pub(crate) fn push_front(&mut self, val: T::Handle) {
+        let ptr = T::to_raw(val);
 
-        entry.as_mut().next = self.head;
-        entry.as_mut().prev = None;
+        unsafe {
+            T::pointers(ptr).as_mut().next = self.head;
+            T::pointers(ptr).as_mut().prev = None;
 
-        if let Some(head) = &mut self.head {
-            head.as_mut().prev = Some(entry);
-        }
+            if let Some(head) = self.head {
+                T::pointers(head).as_mut().prev = Some(ptr);
+            }
 
-        self.head = Some(entry);
+            self.head = Some(ptr);
 
-        if self.tail.is_none() {
-            self.tail = Some(entry);
+            if self.tail.is_none() {
+                self.tail = Some(ptr);
+            }
         }
     }
 
-    /// Removes the first element and returns it, or `None` if the list is empty.
-    ///
-    /// The function is safe as the lifetime of the entry is bound to `&mut
-    /// self`.
-    pub(crate) fn pop_back(&mut self) -> Option<Pin<&mut T>> {
+    /// Removes the last element from a list and returns it, or None if it is
+    /// empty.
+    pub(crate) fn pop_back(&mut self) -> Option<T::Handle> {
         unsafe {
-            let mut last = self.tail?;
-            self.tail = last.as_ref().prev;
+            let last = self.tail?;
+            self.tail = T::pointers(last).as_ref().prev;
 
-            if let Some(mut prev) = last.as_mut().prev {
-                prev.as_mut().next = None;
+            if let Some(prev) = T::pointers(last).as_ref().prev {
+                T::pointers(prev).as_mut().next = None;
             } else {
                 self.head = None
             }
 
-            last.as_mut().prev = None;
-            last.as_mut().next = None;
-
-            let val = &mut *last.as_mut().data.get();
+            T::pointers(last).as_mut().prev = None;
+            T::pointers(last).as_mut().next = None;
 
-            Some(Pin::new_unchecked(val))
+            Some(T::from_raw(last))
```

Comment:
nit: similarly to my above comment, I think the `unsafe` block could be made smaller here...

Reply:
Same response :)